text_generation/causal_lm/cpp/speculative_decoding_lm.cpp

```diff
@@ -0,0 +1,272 @@
+    OPENVINO_ASSERT(vocab_size == main_model.get_tensor("logits").get_shape().back(), "vocab size should be the same for the both models");
+       
+    // logits shape is [BATCH_SIZE, input_ids.size, vocab_size]
+    auto logits = main_model.get_tensor("logits");
+    auto data_logits = logits.data<float>() + (seq_len - 1) * vocab_size;
+    int64_t next_token = std::max_element(data_logits, data_logits + vocab_size) - data_logits;
+    
+    // the first token which is fed to both draft and main netwoks on each iteration
+    auto first_token = next_token;
+    text_streamer.put(next_token);
+    
+    // run K infer requests on draft model and get next K prediction tokens on each iteration
+    uint64_t K = 5;
+    std::vector<int64_t> draft_tokens;
+
+    // The draft model predicts tokens one by one in an auto-regressive manner, draft_input_ids length should be 1.
+    draft_input_ids.set_shape({BATCH_SIZE, 1});
+    draft_position_ids.set_shape({BATCH_SIZE, 1});
+
+    /*
+    Speculative decoding works the following way.
+    The draft model predicts the next K tokens one by one. 
+    These K tokens are then fed to the main model in a single inference request.
+    We go through each predicted token, and if a difference is detected, we stop 
+    and keep the last token predicted by the main network.
+    Then the draft model gets the latest main prediction and again tries to predict 
+    the next K tokens, repeating the cycle.
+    
+    This approach reduces the need for multiple infer requests 
+    to the main model, enhancing performance. For instance, in more predictable 
+    parts of text generation, the draft model can, in best-case scenarios, 
+    generate the next K tokens that exactly match the target. In that case theу
+    are validated in a single inference request of the main model (which is bigger, 
+    more accurate but slower) instead of running K subsequent requests.
+    */
+    int max_sequence_length = 100;
+    while (next_token != SPECIAL_EOS_TOKEN && seq_len < max_sequence_length) {
+        // infer the K next tokens with draft model
+        for (int i = 0; i < K; ++i) {
+            draft_input_ids.data<int64_t>()[0] = next_token;
+            draft_attention_mask.set_shape({BATCH_SIZE, seq_len + i + 1});
+            std::fill_n(draft_attention_mask.data<int64_t>(), draft_attention_mask.get_size(), 1);
+            draft_position_ids.data<int64_t>()[0] = int64_t(draft_attention_mask.get_size() - 1);
+
+            draft_model.infer();
+
+            auto draft_logits = draft_model.get_tensor("logits").data<float>();
+            int64_t arg_max_token = std::max_element(draft_logits, draft_logits + vocab_size) - draft_logits;
+            next_token = arg_max_token;
+            draft_tokens.emplace_back(arg_max_token);
+        }
+
+        // For the main network, K tokens will be fed at once in a single infer request.
+        input_ids.set_shape({BATCH_SIZE, K});
```

Comment:
looks like by this time we actually processed `K + 1` tokens (one before loop, K inside while loop)

Reply:
on the draft model yes, when we get here for the first time`K+1` tokens have been processed, but on the target only one.   

But anyways, number of tokens processed do not affect this number, it's defined only by number of tokens we want to process for the future. So what is your concern here?